bee-message/src/address/ed25519.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{error::ValidationError, signature::Ed25519Signature};
+
+use bee_packable::Packable;
+
+use crypto::{
+    hashes::{blake2b::Blake2b256, Digest},
+    signatures::ed25519::{PublicKey, Signature},
+};
+
+use core::{convert::TryInto, str::FromStr};
+
+/// The number of bytes in an Ed25519 address.
+/// See <https://en.wikipedia.org/wiki/EdDSA#Ed25519> for more information.
+pub const ED25519_ADDRESS_LENGTH: usize = 32;
+
+/// An Ed25519 address.
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Packable)]
+pub struct Ed25519Address([u8; ED25519_ADDRESS_LENGTH]);
+
+#[allow(clippy::len_without_is_empty)]
+impl Ed25519Address {
+    /// The address kind of an Ed25519 address.
+    pub const KIND: u8 = 0;
+
+    /// Creates a new Ed25519 address.
+    pub fn new(address: [u8; ED25519_ADDRESS_LENGTH]) -> Self {
+        address.into()
+    }
+
+    /// Returns the length of an Ed25519 address.
+    pub fn len(&self) -> usize {
+        ED25519_ADDRESS_LENGTH
+    }
+
+    /// Verifies a [`Ed25519Signature`] for a message against the [`Ed25519Address`].
+    pub fn verify(&self, msg: &[u8], signature: &Ed25519Signature) -> Result<(), ValidationError> {
+        let address = Blake2b256::digest(signature.public_key());
+
+        if self.0 != *address {
+            return Err(ValidationError::SignaturePublicKeyMismatch(
+                hex::encode(self.0),
+                hex::encode(address),
+            ));
+        }
+
+        if !PublicKey::from_compressed_bytes(*signature.public_key())?
+            // This unwrap is fine as the length of the signature has already been verified at construction.
+            .verify(&Signature::from_bytes(signature.signature().try_into().unwrap()), msg)
```

Comment:
If you want to be very evil you can cast some raw pointers to avoid the `try_into` runtime checks.

Reply:
Maybe a bit too evil :trollface: